### PR TITLE
[GSoC] Added `DistributionType` to Experiment API

### LINKED ISSUE #2374: [GSOC] Project 8: Support various `Parameter Distribution` in Katib
## Motivation
Currently, [Katib](https://github.com/kubeflow/katib) is limited to supporting only uniform distribution for integer, float, and categorical hyperparameters. By introducing additional distributions, Katib will become more flexible and powerful in conducting hyperparameter optimization tasks.
A Data Scientist requires Katib to support multiple hyperparameter distributions, such as log-uniform, normal, and log-normal, in addition to the existing uniform distribution. For instance, learning rates often benefit from a log-uniform distribution because small values can significantly impact performance. Similarly, normal distributions are useful for parameters that are expected to vary around a central value.
## Goal
The project aims to enhance Katib's Experiment APIs by introducing support for various hyperparameter distributions beyond the current uniform distribution. 
Currently, the Experiment APIs support `parameterType` and `feasibleSpace`. To provide greater flexibility, we expanded the `feasibleSpace` to include a `Distribution` field, allowing users to specify the type of distribution for each parameter.

This Distribution field is designed to be mutually exclusive, meaning it can be left empty, enabling users to tune hyperparameters without specifying a distribution if desired.

### Preparation
- [x] Converted the GSoC proposal into a KEP: #2334
### Development
- [x] Added `Distribution` field to the `FeasibleSpace` object: #2377 
- [x] Updated gRPC APIs to include the `Distribution` field: #2397
- [x] Modified Suggestion client to handle the new Distribution field: #2403 
- [x] Updated validation webhook to support both ParameterType and Distribution: #2404 
- [x] Implemented logic updates for handling new distributions in `hyperopt`suggestion service: #2412
- [ ] Implemented logic updates for handling new distributions in `optuna`suggestion service
- [ ] Implemented logic updates for handling new distributions in `goptuna`suggestion service
### Documentation and Ongoing Work
- [x] Write e2e tests for `normal` and `lognormal` distributions in `hyperopt`: #2412 
- [ ] Extend the Python SDK to include support for the `Distribution` field.
- [ ] Update Katib documentation to reflect new API changes and usage guidelines.
- [ ] Create examples demonstrating the use of different parameter distributions.

## What I learnt during the project
- One of the key hurdles was integrating multiple new distributions into the Katib Experiment APIs while ensuring backward compatibility. Modifying the `feasibleSpace` to include a `Distribution` field required careful consideration of API design and how different suggestion services, like Optuna and Hyperopt, would handle these changes.

- Effective communication is key. The weekly syncs with @andreyvelich and @tenzen-y were extremely valuable, and having a clear agenda beforehand helped me get my questions answered efficiently and stay on track.

- I learned the importance of end-to-end (E2E) tests. By implementing these tests, I gained valuable insights into how the different components of Katib interact and how critical it is to verify that new features, like the distribution support.

## In the End
I am grateful to Google for the opportunity to work on this project and to the Kubeflow community for their support and guidance. Special thanks to my project mentors, @andreyvelich, @tenzen-y, and @terrytangyuan  for your invaluable insights and encouragement. Your mentorship was pivotal in achieving the project's goals.

I hope the enhancements made will significantly benefit users and advance Katib’s capabilities in hyperparameter optimization. Looking forward to continue to contribute to Kubeflow.

### PR DESCRIPTION
**Pull request by shashank-iitbhu**
**What this PR does / why we need it**:
- Added `DistributionType` to `ParameterSpec` in `pkg/apis/controller/experiments/v1beta1/experiment_types.go`.
- Defined `DistributionType`.

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2374 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Comment by shashank-iitbhu** [comment]
`ParameterType` and `DistributionType` can be made exclusive using this logic
```go
// Check that only one of ParameterType or DistributionType is specified
        if (param.ParameterType != "" && param.DistributionType != "") ||
            (param.ParameterType == "" && param.DistributionType == "") {
            allErrs = append(allErrs, field.Invalid(parametersPath.Index(i).Child("parameterType"),
                param.ParameterType, "either parameterType or distributionType must be specified, but not both"))
        }
``` 
at
https://github.com/kubeflow/katib/blob/154a85b7408a827c35150874c1197ec910072bab/pkg/webhook/v1beta1/experiment/validator/validator.go#L254-L256

File: pkg/apis/controller/experiments/v1beta1/experiment_types.go [hunk 1]
```diff
@@ -191,9 +191,10 @@ const (
 )
 
 type ParameterSpec struct {
-	Name          string        `json:"name,omitempty"`
-	ParameterType ParameterType `json:"parameterType,omitempty"`
-	FeasibleSpace FeasibleSpace `json:"feasibleSpace,omitempty"`
+	Name             string           `json:"name,omitempty"`
+	ParameterType    ParameterType    `json:"parameterType,omitempty"`
+	DistributionType DistributionType `json:"distributionType,omitempty"`
```
Comment (andreyvelich) [inline pkg/apis/controller/experiments/v1beta1/experiment_types.go:196]: Based on our last discussion, do you want to move the `Distribution` under `FeasibleSpace` ?

Comment (shashank-iitbhu) [reply pkg/apis/controller/experiments/v1beta1/experiment_types.go:196]: Yes, will update this PR before next sync. Out of station right now.

File: pkg/apis/controller/experiments/v1beta1/experiment_types.go [hunk 2]
```diff
@@ -207,12 +207,23 @@ const (
...
 }
 
+type Distribution string
+
+const (
+	DistributionUniform     Distribution = "uniform"
+	DistributionLogUniform  Distribution = "logUniform"
+	DistributionNormal      Distribution = "normal"
+	DistributionLogNormal   Distribution = "logNormal"
+	DistributionCategorical Distribution = "categorical"
```
Comment (andreyvelich) [inline pkg/apis/controller/experiments/v1beta1/experiment_types.go:224]: We should remove this.

File: pkg/apis/controller/experiments/v1beta1/experiment_types.go (hunk 1 above)
Comment (shashank-iitbhu) [reply pkg/apis/controller/experiments/v1beta1/experiment_types.go]: Done

File: pkg/apis/controller/experiments/v1beta1/experiment_types.go (hunk 2 above)
Comment (shashank-iitbhu) [reply pkg/apis/controller/experiments/v1beta1/experiment_types.go]: Yes, removed Categorical from `Distribution` type.

**Review by tenzen-y — COMMENTED** [review]
Thank you!
/lgtm
/approve

**Pull request merged**